### ISSUE TITLE
Fails to build with Clang-13 and c++20 standard (unverified)

### ISSUE THREAD

**Issue by vnepogodin** [simdjson/simdjson#1776]
**Describe the bug**
I want to statically link simdjson to my app, which uses c++20 standard.
Compiling with GCC is fine, but Clang is fails due to unfinished work on c++20 ranges library.

```bash
[ 50%] Building CXX object CMakeFiles/simdjson.dir/src/simdjson.cpp.o
In file included from /home/vl/Downloads/simdjson/src/simdjson.cpp:1:
In file included from /home/vl/Downloads/simdjson/include/simdjson.h:37:
In file included from /home/vl/Downloads/simdjson/include/simdjson/dom.h:9:
In file included from /home/vl/Downloads/simdjson/include/simdjson/dom/array.h:170:
/sbin/../lib64/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/ranges:3392:19: error: missing 'typename' prior to dependent type name 'iterator_traits<iterator_t<_Base>>::iterator_category'
            using _Cat = iterator_traits<iterator_t<_Base>>::iterator_category;
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**To Reproduce**
```bash
git clone https://github.com/simdjson/simdjson.git
cd simdjson
CXX=clang++ cmake -S . -B build -DCMAKE_CXX_STANDARD="20"
cmake --build build
```

The issue goes away by removing include at `include/simdjson/dom/object.h:260`
and at `include/simdjson/dom/array.h:170`.

Or using `CXXFLAGS="$CXXFLAGS -stdlib=libc++` instead.

**simdjson release**
`master branch`
commit hash: https://github.com/simdjson/simdjson/commit/3a93e45dd166d8e614a33b0ba4da39c1a01660b7

**Configuration (please complete the following information if relevant)**
 - OS: Arch Linux
 - Compiler: clang version 13.0.0

**Indicate whether you are willing or able to provide a bug fix as a pull request**
No

**Comment by lemire** [comment]
@llllllllll Would you be willing to assess this issue?

**Comment by lemire** [comment]
I am trying to understand the issue,

Line 170 is the include there...

```C++
#if defined(__cpp_lib_ranges)
#include <ranges>
...
```

**Comment by lemire** [comment]
I cannot verify the issue.

```
$ clang++ --version
Ubuntu clang version 13.0.0-9
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/bin

$ CXX=clang++ cmake  -DCMAKE_CXX_STANDARD="20" -B build20
$ cmake --build build

$ cmake --build build20
[ 50%] Building CXX object CMakeFiles/simdjson.dir/src/simdjson.cpp.o
In file included from /Users/lemire/CVS/github/simdjson/src/simdjson.cpp:1:
In file included from /Users/lemire/CVS/github/simdjson/include/simdjson.h:38:
In file included from /Users/lemire/CVS/github/simdjson/include/simdjson/builtin.h:4:
In file included from /Users/lemire/CVS/github/simdjson/include/simdjson/implementations.h:50:
In file included from /Users/lemire/CVS/github/simdjson/include/simdjson/arm64.h:25:
/Users/lemire/CVS/github/simdjson/include/simdjson/arm64/simd.h:510:25: warning: ISO C++20 considers use of overloaded operator '==' (with operand types 'const simd8<unsigned char>' and 'const simd8<unsigned char>') to be ambiguous despite there being a unique best viable function [-Wambiguous-reversed-operator]
        this->chunks[0] == mask,
        ~~~~~~~~~~~~~~~ ^  ~~~~
/Users/lemire/CVS/github/simdjson/src/generic/stage1/json_string_scanner.h:111:33: note: in instantiation of member function 'simdjson::arm64::(anonymous namespace)::simd::simd8x64<unsigned char>::eq' requested here
  const uint64_t backslash = in.eq('\\');
                                ^
/Users/lemire/CVS/github/simdjson/include/simdjson/arm64/simd.h:135:33: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
    simdjson_really_inline Mask operator==(const simd8<T> other) const { return vceqq_u8(*this, other); }
                                ^
/Users/lemire/CVS/github/simdjson/include/simdjson/arm64/simd.h:511:25: warning: ISO C++20 considers use of overloaded operator '==' (with operand types 'const simd8<unsigned char>' and 'const simd8<unsigned char>') to be ambiguous despite there being a unique best viable function [-Wambiguous-reversed-operator]
        this->chunks[1] == mask,
        ~~~~~~~~~~~~~~~ ^  ~~~~
/Users/lemire/CVS/github/simdjson/include/simdjson/arm64/simd.h:135:33: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
    simdjson_really_inline Mask operator==(const simd8<T> other) const { return vceqq_u8(*this, other); }
                                ^
/Users/lemire/CVS/github/simdjson/include/simdjson/arm64/simd.h:512:25: warning: ISO C++20 considers use of overloaded operator '==' (with operand types 'const simd8<unsigned char>' and 'const simd8<unsigned char>') to be ambiguous despite there being a unique best viable function [-Wambiguous-reversed-operator]
        this->chunks[2] == mask,
        ~~~~~~~~~~~~~~~ ^  ~~~~
/Users/lemire/CVS/github/simdjson/include/simdjson/arm64/simd.h:135:33: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
    simdjson_really_inline Mask operator==(const simd8<T> other) const { return vceqq_u8(*this, other); }
                                ^
/Users/lemire/CVS/github/simdjson/include/simdjson/arm64/simd.h:513:25: warning: ISO C++20 considers use of overloaded operator '==' (with operand types 'const simd8<unsigned char>' and 'const simd8<unsigned char>') to be ambiguous despite there being a unique best viable function [-Wambiguous-reversed-operator]
        this->chunks[3] == mask
        ~~~~~~~~~~~~~~~ ^  ~~~~
/Users/lemire/CVS/github/simdjson/include/simdjson/arm64/simd.h:135:33: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
    simdjson_really_inline Mask operator==(const simd8<T> other) const { return vceqq_u8(*this, other); }
                                ^
4 warnings generated.
[100%] Linking CXX static library libsimdjson.a
[100%] Built target simdjson
```

**Comment by lemire** [comment]
Given that you report an error that happens at the point where we include a standard file, I suspect that you may have a configuration issue that is not specific to simdjson.

If I am wrong, please elaborate. Including " <ranges>" should be fine.

**Comment by furkanusta** [comment]
It seems that libstdc++ has a missing 'typename' in 11.1.0 https://github.com/gcc-mirror/gcc/blob/50bc9185c2821350f0b785d6e23a6e9dcde58466/libstdc%2B%2B-v3/include/std/ranges#L3392

and it seems to be added in 11.2.0
https://github.com/gcc-mirror/gcc/blob/7ca388565af176bd4efd4f8db1e5e9e11e98ef45/libstdc%2B%2B-v3/include/std/ranges#L3600

Not sure if there is anything that can be done in simdjson's part to circumvent the issue.

**Comment by vnepogodin** [comment]
https://github.com/ericniebler/range-v3/pull/1635

**Comment by vnepogodin** [comment]
> It seems that libstdc++ has a missing 'typename' in 11.1.0 https://github.com/gcc-mirror/gcc/blob/50bc9185c2821350f0b785d6e23a6e9dcde58466/libstdc%2B%2B-v3/include/std/ranges#L3392
> 
> and it seems to be added in 11.2.0 https://github.com/gcc-mirror/gcc/blob/7ca388565af176bd4efd4f8db1e5e9e11e98ef45/libstdc%2B%2B-v3/include/std/ranges#L3600
> 
> Not sure if there is anything that can be done in simdjson's part to circumvent the issue.

yeah on Arch Linux which I'm using. GCC 11.1.0 is latest
But its GCC working as I said above

**Comment by lemire** [comment]
I am going to close this issue. As far as I can tell, the issue is that on some unpatched systems, enabling C++20 will not work because the compiler can't even compile the standard C++20 headers.

My recommendation is that users upgrade their systems, or do without C++20.

To be clear, I don't think a C++20 platform that fails when you include this line...

```C++
#include <ranges>
```

is a serious platform for C++20 work. Especially because we guard it with...

```C++
#if defined(__cpp_lib_ranges)
```

It is possible that I missed something, but I think we are entirely in the right. The defect is with the toolchain.

If someone has a constructive solution and they want to prepare a reasonable pull request... then we shall review it. However, I am not too keen on patching something that is clearly a defect of the toolchain, even more so if the defect was corrected.

**Comment by llllllllll** [comment]
I think the issue here is that the standard relaxed the requirement to use the `typename` disambiguator in many contexts where a type was required and a value was invalid, and ranges is a libstdc++ feature which is only valid in C++20 so it was written without the disambiguators. gcc has supported this feature since gcc-9; however, clang still doesn't support the feature. As @lemire said, this is not an issue with ranges, but rather an issue mixing clang with your current version of libstdc++.

The new feature is this paper: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0634r3.html
clang's feature support matrix is here: https://clang.llvm.org/cxx_status.html, where they refer to this as "typename optional in more contexts".
gcc's feature support matrix is here: https://gcc.gnu.org/projects/cxx-status.html, where they refer to this as "Down with typename!" (the name of the paper).

**Comment by vnepogodin** [comment]
Yes @llllllllll you right.
I suggested above that simdjson can use solution from range-v3 library described  in their PR: 
https://github.com/ericniebler/range-v3/pull/1635

**Comment by lemire** [comment]
@vnepogodin You are linking to an implementation of the C++20 range feature with a fix for this correction.

I am quite certain that @ldionne (the author of the patch) does not mean for every user of the C++20 range feature to roll out their own patch to the standard library.

In any case, my call for a pull request remains.

**Comment by lemire** [comment]
That is, if you think that my analysis is wrong, and we can fix the issue, then please propose a fix.

**Comment by llllllllll** [comment]
The reason for that PR on range-v3 is that it _is_ an implementation of ranges, it doesn't just use ranges

**Comment by lemire** [comment]
> The reason for that PR on range-v3 is that it is an implementation of ranges, it doesn't just use ranges

That is my understanding.

**Comment by ldionne** [comment]
I don't see the relationship between @vnepogodin 's original bug report, my fix of ericniebler/range-v3#1635 and simdjson.

The original bug report here is that recent libstdc++ doesn't work with recent Clang due to Clang not implementing the relaxed requirements for `typename` appearance. That is a valid complaint, but it is not related to simdjson.

The bug fixed by myself in ericniebler/range-v3#1635 has to do with using `range-v3` alongside a recent libc++. `range-v3` was using some internal details of libc++ and that broke when we changed those implementation details. This is not related to simdjson either.

Both these issues are unrelated to each other, and both are unrelated to simdjson -- I think there is some confusion here. Long story short, I think this should stay closed and there's nothing actionable for anyone to do, except either for Clang to implement p0634r3 or for libstdc++ not to assume that the compiler implements it. @vnepogodin if you want to report these issues, please do it in the GCC bug tracker or in the Clang bug tracker, depending on what you want to report.
